framework/src/tinyalsa/tinyalsa.c

```diff
@@ -1292,82 +1365,95 @@ int pcm_wait(struct pcm *pcm, int timeout)
-		/* We dont have any empty buffers. wait for deque message from kernel */
-		if (timeout > 0) {
-			clock_gettime(CLOCK_REALTIME, &st_time);
-			st_time.tv_nsec += timeout * MILLI_TO_NANO;
-			size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
-		} else {
-			size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
-		}
 
+	/* If there are some messages in the queue, empty them */
+	while (1) {
+		clock_gettime(CLOCK_REALTIME, &st_time);
+		size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
 		if (size != sizeof(msg)) {
-			if (errno == ETIMEDOUT) {
-				oops(pcm, errno, "TIMEOUT while watiting for deque message from kernel");
-				return 0;
-			} else {
-				return oops(pcm, errno, "Interrupted while waiting for deque message from kernel");
-			}
+			break;
 		}
+
+		/* If you find a dequeue message in the queue,
+		update the apb flags and curbyte and increment the count*/
 		if (msg.msgId == AUDIO_MSG_DEQUEUE) {
 			apb = (struct ap_buffer_s *)msg.u.pPtr;
-			pcm->next_buf = apb;
-			pcm->next_size = apb->nbytes;
-			pcm->buf_ptr--;
-			return 1;
+			apb->flags &= ~AUDIO_APB_OUTPUT_ENQUEUED;
+			apb->curbyte = 0;
+			count++;
+		}
+	}
+
+	if (count > 0) {
+		return 1;
+	}
+
+	/* If there were no buffers in the queue, wait for codec to put a buffer on the queue */
+	if (timeout > 0) {
+		/* Use the timeout given by application */
+		clock_gettime(CLOCK_REALTIME, &st_time);
+		st_time.tv_nsec += timeout * MILLI_TO_NANO;
+		size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
+	} else {
+		/* Application did not give a timeout value
+		The behavior here depends on whether mq was opened with O_NONBLOCK
+		This might block until a msg becomes available on the queue */
+		size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
+	}
+
+	if (size != sizeof(msg)) {
+		if (errno == ETIMEDOUT) {
+			oops(pcm, errno, "TIMEOUT while watiting for deque message from kernel");
+			return 0;
 		} else {
-			return oops(pcm, EINTR, "Recieved unexpected msg (id = %d) while waiting for deque message from kernel", msg.msgId);
+			return oops(pcm, errno, "Interrupted while waiting for deque message from kernel");
 		}
 	}
+	if (msg.msgId == AUDIO_MSG_DEQUEUE) {
+		/* If you find a dequeue message in the queue,
+		update the apb flags and curbyte and increment the count*/
+		apb = (struct ap_buffer_s *)msg.u.pPtr;
+		apb->flags &= ~AUDIO_APB_OUTPUT_ENQUEUED;
+		apb->curbyte = 0;
+		return 1;
+	}
+
+	return oops(pcm, EINTR, "Recieved unexpected msg (id = %d) while waiting for deque message from kernel", msg.msgId);
 }
 
 int pcm_mmap_transfer(struct pcm *pcm, const void *buffer, unsigned int bytes)
 {
 	int err = 0, frames, avail;
 	unsigned int offset = 0, count;
 
+	/* No data to be transferred */
 	if (bytes == 0) {
 		return 0;
 	}
 
-	count = pcm_bytes_to_frames(pcm, bytes);
+	/* Start the PCM, if required */
+	if (pcm->flags & PCM_IN && !pcm->running) {
```

Comment:
Should PCM_MMAP be checked here as we check for PCM_IN in pcm_readi

Reply:
We are checking in pcm_mmap_read and pcm_mmap_write, and this API is not exposed to user. 
So, no need to check here.